pyi.py

```diff
@@ -278,33 +299,10 @@ class PyiVisitor(ast.NodeVisitor):
     typevarlike_defs: dict[TypeVarInfo, ast.Assign] = field(default_factory=dict)
     # Mapping of each name in the file to the no. of occurrences
     all_name_occurrences: Counter[str] = field(default_factory=Counter)
-    _class_nesting: int = 0
-    _function_nesting: int = 0
-    _allow_string_literals: int = 0
-
-    @contextmanager
-    def allow_string_literals(self) -> Iterator[None]:
-        """Context manager that indicates that string literals should be allowed."""
-        self._allow_string_literals += 1
-        try:
-            yield
-        finally:
-            self._allow_string_literals -= 1
-
-    @property
-    def string_literals_allowed(self) -> bool:
-        """Determine whether string literals should currently be allowed."""
-        return bool(self._allow_string_literals)
-
-    @property
-    def in_function(self) -> bool:
-        """Determine whether we are inside a `def` statement"""
-        return bool(self._function_nesting)
 
-    @property
-    def in_class(self) -> bool:
-        """Determine whether we are inside a `class` statement"""
-        return bool(self._class_nesting)
+    string_literals_allowed: AttributeManager = AttributeManager()
+    in_function: AttributeManager = AttributeManager()
+    in_class: AttributeManager = AttributeManager()
```

Comment:
Do these really get created instance-specifically? I think you need to use `dataclasses.field`.

Reply:
Oh, yup, you're right.